doc/design/cluster_train/save_model.md

```diff
@@ -0,0 +1,100 @@
+# Design Doc: Save Model
+
+## Overview
+
+The model is the output of the training process. There are two
+ways from which user can obtain a model:
+
+- Save model triggered by user code: user code asks PaddlePaddle to
+  save a model.
+- Convert model from the snapshot: model being converted from
+  pservers' periodic snapshot. In this way, the user can cancel a job
+  at any time, and still have a relatively fresh model (we snapshot
+  around every 5 minutes).
+
+### Save Model Triggered by User Code
+
+Both trainers and pservers have access to the model. So the model can
+be saved from a trainer or pservers. We need to decide on where the
+model is saved from.
+
+#### Dense Model vs. Sparse Model
+
+There are two types of model: dense and sparse model (when the
+parameter is configured to be sparse). Pservers always jointly have
+the entire model at any given time. Trainers only have the entire
+dense model, but only have a fraction of the sparse model at any given
+time.
+
+#### Pservers Saving Model
```

Comment:
@typhoonzero @dzhwinter 
> saving the snapshot at pserver side is needed for recovering the state of pservers. A pserver snapshot should contain not only parameters but also some status like optimizer internels.

Agree!

> Saving snapshot by pserver can be triggered by call Save() RPC call from trainers to pservers.

I think we can add that if we find it necessary later. But for the first version I am inclining to just let the pservers save periodically.

> Trainers can save models by parameter.to_tar() in event_handlers.

I think trainer need Python function called `save_model(save_dir)`, in the Python function it will first ask trainer client if it is elected to save the model, save the model if elected. Otherwise every trainer will try to save the model, putting too much burden to the distributed FS.

> Pserver side "snapshot" will only be used for pserver recovery, but trainer side saved model can be used for inference.

Agree!